File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    visited = [0 for _ in range(len(cards))]
    group = []
    
    for i in range(len(cards)):
        cnt = 0
        while visited[i] == 0:
            visited[i] = 1
            i = cards[i] - 1
            cnt += 1
        group.append(cnt)
            
    group.sort(reverse = True)
    return group[0] * group[1]

# 그룹이 몇갠지 파악 -> 그룹 속 카드 많은 두개 택 
# 싸이클 여부 

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    public int solution(int[] cards) {
        boolean[] visited = new boolean[cards.length];
        ArrayList<Integer> group = new ArrayList<>();
        
        for (int i = 0; i< cards.length; i++) {
            if (!visited[i]) {
                int cnt = 0;
                int index = i;
                while (!visited[index]) {
                    visited[index] = true;
                    index = cards[index] - 1;
                    cnt += 1;
                }
                group.add(cnt);
            }
        }
        Collections.sort(group, Collections.reverseOrder());
        
        if (group.size() <= 1) {
            return 0;
        }
        return group.get(0) * group.get(1);
        
    }
}